### ISSUE TITLE
[NTI] Module scope variable type not known inside function

### ISSUE THREAD

**Issue by syranide** [google/closure-compiler#2774]
Put the below code anywhere in the top-level scope of any file (in this case, `src/index.js`).

```js
let inputElement = /** @type {!HTMLInputElement} */ (document.createElement('input'));
function test() {
  inputElement.value = '';
}
```

Generates the below warning:

```
src/index.js:3: WARNING - This GETPROP expression has the unknown type.
  inputElement.value = '';
  ^^^^^^^^^^^^^^^^^^
```

However all of the following trivially tweaked examples do work correctly and do not generate a warning:

```js
/** @type {!HTMLInputElement} */
let inputElement = /** @type {!HTMLInputElement} */ (document.createElement('input'));
function test() {
  inputElement.value = '';
}
```
```js
let inputElement = /** @type {!HTMLInputElement} */ (document.createElement('input'));
function test() {
  let localInputElement = inputElement;
  localInputElement.value = '';
}
```
```js
let inputElement = /** @type {!HTMLInputElement} */ (document.createElement('input'));
inputElement.value = '';
```

This seems illogical to me, especially given that simply reassigning it to a local variable makes it work.

----

Compiled with (npm-version 20171112.0.0):

> java -XX:+TieredCompilation -jar node_modules/google-closure-compiler/compiler.jar -O ADVANCED --checks_only --js \"src/**.js\" --entry_point \"src/index\" --language_out ECMASCRIPT5_STRICT --new_type_inf --dependency_mode STRICT --env BROWSER --module_resolution NODE --warning_level=VERBOSE --jscomp_warning=newCheckTypes --jscomp_warning=reportUnknownTypes --jscomp_warning=extraRequire

**Comment by blickly** [comment]
Probably @dimvar is the best person to look at this.

**Comment by dimvar** [comment]
It's good to annotate variables that are used across scope, and not rely on inference, especially if you want no unknown types (see [this](https://youtu.be/MuC8I1JBKv0?t=25m27s) also).

**Comment by syranide** [comment]
@dimvar Fair enough and I agree. However, that explanation doesn't really seem to agree with my second "working example", where just by assigning locally the type is inferred correctly (I assume), unless that behavior is unintentional.
